interpreter/cling/lib/Interpreter/Interpreter.cpp

```diff
@@ -666,6 +695,62 @@ namespace cling {
   }
 
   Interpreter::CompilationResult
+  Interpreter::codeComplete(const std::string& line, size_t& cursor,
+                            std::vector<std::string>& completions) const {
+
+    const char * const argV = "cling";
+    std::string llvmdir = this->getCI()->getHeaderSearchOpts().ResourceDir;
+    std::string extra_part = "/lib/clang/3.9.0";
+    std::string::size_type i = llvmdir.find(extra_part);
+    if (i != std::string::npos)
+      llvmdir.erase(i, extra_part.length());
+
+    cling::Interpreter childInterpreter(*this, 1, &argV, llvmdir.c_str());
+
+    auto childCI = childInterpreter.getCI();
+    clang::Sema &childSemaRef = childCI->getSema();
+
+    // Create the CodeCompleteConsumer with InterpreterCallbacks
+    // from the parent interpreter and set the consumer for the child
+    // interpreter.
+    ClingCodeCompleteConsumer* consumer = new ClingCodeCompleteConsumer(
+              this->getCI()->getFrontendOpts().CodeCompleteOpts, llvm::errs(),
+              completions);
+    // Child interpreter CI will own consumer!
+    childCI->setCodeCompletionConsumer(consumer);
+    childSemaRef.CodeCompleter = consumer;
+
+    // Ignore diagnostics when we tab complete.
+    // This is because we get redefinition errors due to the import of the decls.
+    clang::IgnoringDiagConsumer* ignoringDiagConsumer =
+                                            new clang::IgnoringDiagConsumer();
+    childSemaRef.getDiagnostics().setClient(ignoringDiagConsumer, true);
+
+    auto ownerDiagConsumer =
+                        this->getCI()->getSema().getDiagnostics().takeClient();
+    auto clientDiagConsumer =
+                        this->getCI()->getSema().getDiagnostics().getClient();
+    this->getCI()->getSema().getDiagnostics().setClient(
+                                                  ignoringDiagConsumer, false);
+
+    // The child will desirialize decls from *this. We need a transaction RAII.
+    PushTransactionRAII RAII(this);
+
+    // Triger the code completion.
+    childInterpreter.CodeCompleteInternal(line, cursor);
+
+    // Restore the original diagnostics client for parent interpreter.
+    this->getCI()->getSema().getDiagnostics().setClient(
+                    clientDiagConsumer, ownerDiagConsumer.release() != nullptr);
+
+    // FIX-ME : Change it in the Incremental Parser
+    // It does not work by call unload in IncrementalParser, might be to early.
+    childInterpreter.unload(1);
```

Comment:
Can't we roll back the transaction in the `IncrementalParser`, as we do in error cases?